Built-in self-updates for Linux, macOS, Windows

vadi2 opened the pull request:
Still a WIP and a lot of work to do, but it can update the Linux version already. Comments as I work on this as welcome.

keneanung commented:
The way the SDK gets added to the project gives me headaches. I will contact the maintainer and work with them to get a clean way to include the project.

Cmake is also still missing.

vadi2 commented:
Yeah I've requested cmake support from them.

On Tue, 10 Oct 2017, 10:30 pm keneanung, <notifications@github.com> wrote:

> The way the SDK gets added to the project gives me headaches. I will
> contact the maintainer and work with them to get a clean way to include the
> project.
>
> Cmake is also still missing.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/1318#issuecomment-335598203>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGxjCAqPkGdx-Z1ZL3ypceuo8fqE8huks5sq9PngaJpZM4P0J55>
> .
>

vadi2 commented:
I've got the code setup to expect `Mudlet.AppImage` to come out from the tar archive, which means that if we decide to rename it in the future, the updater will break.

How should we go about this? I think a reasonable expectation would be to:

1. make it handle both direct AppImages and tarred ones, in case we don't have to tar them anymore
1. make it handle anything *.AppImage case-insensitive in case we rename the binary, e.g start adding version strings

That seems good enough for future-proofing for me.

vadi2 commented:
I've added an option to disable automatic updates in settings now, so you can have it both ways - automated (default) and manual - shows a dialog.

vadi2 commented:
https://bitbucket.org/pentacent/dblsqd-sdk-qt/commits/f61a0652855508914b8d8f72c0262f03de539c00 @keneanung was that you?

keneanung commented:
That was the reaction to my mail. I'm not completely happy with it yet, but it show the direction they want to take and I can create PRs against that.

vadi2 commented:
Linux WIP of the autoupdater: https://transfer.sh/Jyxns/Mudlet-3.4.0-testing-PR1318-86185d6-linux-x64.AppImage.tar

All it can do right now is autoupdate down to the 3.4.0 version (which doesn't have the autoupdater), or if you go into preferences and disable autoupdates, it'll show an update dialog when you quit Mudlet. The update dialog can't do updates yet, it'll just download and open the file.

It's something to try anyway. The things I've still gotta do:

1) add a "whats new" dialog when you first launch
2) add a force version check to preferences so you can do it ondemand
3) make mac, windows updates work

SlySven commented:
Re:

> ... make it handle anything *.AppImage case-insensitive in case we rename the binary, e.g start adding version strings ...

that will have to case insensitive to hand some macOs file-system issues won't it?

vadi2 commented:
Not just that, it's also so we aren't tied to spelling it `.AppImage`, `.appimage`, or however else :)

vadi2 commented:
It seems you can't pull off the same easy trick on macOS as on Linux - updating the binary while its open. Will have to use Sparkle or Squirrel, but the difficulty with those is that all official instructions are geared towards Objective C.

Problem 2 is that it has to be signed, which requires an Apple Developer ID at a cost of $99/year.

SlySven commented:
CI failed on Travis for 2 out of 7 case:
+ MacOS-Qt5.9-CMake build (Clang compiler is implicit obviously on THAT OS) - during linking
+ Linux-Qt5.6-GCC-CMake build (but Qt5.9 is okay with same compiler and build system) - in:
>
>     In file included from /home/travis/build/Mudlet/Mudlet/3rdparty/dblsqd/dblsqd/update_dialog.cpp:1:0:
>     /home/travis/build/Mudlet/Mudlet/3rdparty/dblsqd/dblsqd/update_dialog.h:20:46: error: ‘Type’ is not a class or namespace
>     explicit UpdateDialog(Feed* feed, Type = Type::OnUpdateAvailable, QWidget* parent = 0, QSettings* settings = new QSettings());

I had kicked those 2 to try again but they failed again a day later. 😞

We do not check any other Qt5.6 variants (i.e. Clang or QMake options) so cannot tell whether this is a Qt 5.6 specific issue (for the second fail)...

keneanung commented:
Qt 5.6 is a dblsqd defect that was fixed upstream, Vadi just didn't sync yet. The other might be that cmake isn't configured correctly for the sparkle dependency yet.

vadi2 commented:
Tip for getting started on Squirrel.Windows for us: https://github.com/Squirrel/Squirrel.Windows/issues/838

keneanung commented:
There is a tutorial-like thing at http://markwal.github.io/programming/2015/07/28/squirrel-for-windows.html

vadi2 commented:
Welp, this PR is getting huge, and I've only got macOS and Linux updates in here - not Windows ones.

@SlySven @keneanung did you want to review and merge the macOS and Linux integration now, and do the Windows one in a separate PR?

This is on the caveat that we can't release then until Windows updates are ready as well, because releasing an update with a feature that's not available on all platforms is a no-go.

For macOS and Linux, I only have 2 things left to go: 1) make cmake build work with sparkle and 2) integrate dblsqd's automatic checkbox into the one built into Mudlet. Otherwise, it's ready for review.

SlySven commented:
When it comes to merging this is __definitely__ one for "Squash and merge" - together with a commit message that combines the valid parts of all the component commits - otherwise we are not going to be able to see the wood for the trees... 🙂

SlySven commented on src/updater.cpp:
```diff
@@ -0,0 +1,187 @@
+    tar.setProcessChannelMode(QProcess::MergedChannels);
+    // we can assume tar to be present on a Linux system. If it's not, it'd be rather broken.
+    // tar output folder has to end with a slash
+    tar.start("tar", QStringList() << "-xvf" << fileName << "-C" << QStandardPaths::writableLocation(QStandardPaths::TempLocation) + QStringLiteral("/"));
```
You could also assemble the arguments before use (might help if debugging?):

    QStringList tarArguments;
    tarArguments << QLatin1String("-xvf") << fileName << "-C" << QStringLiteral("%1/").arg(QStandardPaths::writableLocation(QStandardPaths::TempLocation));
    tar.setArguments(tarArguments);
    tar.start(QStringLiteral("tar"));

vadi2 replied:
Could, but then it's one line versus four... already debugged this part as well.

SlySven commented on src/updater.cpp:
```diff
@@ -0,0 +1,187 @@
+    // we can assume tar to be present on a Linux system. If it's not, it'd be rather broken.
+    // tar output folder has to end with a slash
+    tar.start("tar", QStringList() << "-xvf" << fileName << "-C" << QStandardPaths::writableLocation(QStandardPaths::TempLocation) + QStringLiteral("/"));
+    if (!tar.waitForFinished()) {
```
Is the default 30 seconds wait (will freeze the main GUI if this is in the main thread) the right amount here?

vadi2 replied:
Yes - because this is never run on the main GUI thread ;)

vadi2 replied:
I've added an assert to ensure it does not.

SlySven commented on src/updater.cpp:
```diff
@@ -0,0 +1,187 @@
+    if (!tar.waitForFinished()) {
+        qWarning() << "Untarring" << fileName << "failed:" << tar.errorString();
+    } else {
+        qWarning() << "Tar output:" << tar.readAll().trimmed();
```
Wouldn't a `qDebug()` be a better option for this "success" message?

vadi2 replied:
Shouldn't spam with a success message at all - will remove this one, thanks.

csmall commented:
Self update messages and features are a pain for all distributions.  Fine to have it there as long as there is a compile time flag to turn it off.

Even if the mudlet binary could update  (which means you are running a mud client as root to get the right permissions and deserve all the punishment that brings) the libraries would be out of kilter.

I know there is different ways of running mudlet and for a statically linked or bring your own libraries it's fine but please have it conditional on behalf of all distributions

vadi2 commented:
Yep I'll add a compile-time flag to turn it off. Btw still waiting on your input on what are the new URLs to use at https://wiki.mudlet.org/w/Updating_PPA

vadi2 commented on src/dlgProfilePreferences.cpp:
```diff
@@ -442,6 +444,13 @@ void dlgProfilePreferences::loadEditorTab()
+    if (mudlet::self()->onDevelopmentVersion()) {
+        // tick the box and make it be untickable as automatic updates are disabled in dev builds
+        checkbox_noAutomaticUpdates->setEnabled(true);
+        checkbox_noAutomaticUpdates->setDisabled(true);
```
Yes, that's what I meant - will fix

vadi2 commented on src/mudlet.cpp:
```diff
@@ -422,7 +423,12 @@ mudlet::mudlet()
+    // hide About->Update on macOS as Sparkle will add its own menu
+    dactionUpdate->setVisible(false);
+#endif
+    connect(dactionUpdate, &QAction::triggered, this, &mudlet::slot_check_manual_update);
```
Done

vadi2 commented on src/mudlet.cpp:
```diff
@@ -152,17 +151,19 @@ mudlet::mudlet()
     setUnifiedTitleAndToolBarOnMac(true);
     setContentsMargins(0, 0, 0, 0);
-    mudlet::debugMode = false;
+    mudlet::debugMode = false;    
```
Fixed

vadi2 commented on src/updater.cpp:
```diff
@@ -0,0 +1,187 @@
+
+#ifdef Q_OS_MACOS
+    setupOnMacOS();
+#endif
```
Improved

vadi2 commented on src/updater.cpp:
```diff
@@ -0,0 +1,187 @@
+}
+#endif
+
+void Updater::setupOnLinux()
```
Went with your preferred version :)

vadi2 commented on src/mudlet.cpp:
```diff
@@ -2053,16 +2059,36 @@ void mudlet::closeEvent(QCloseEvent* event)
+        console->getHost()->saveProfile();
         console->mUserAgreedToCloseConsole = true;
+
+        if (console->getHost()->getName() != QStringLiteral("default_host")) {
```
Good catch - it actually stayed around and crashed.

vadi2 replied:
Fixed.

SlySven commented on src/CMakeLists.txt:
```diff
@@ -381,6 +381,10 @@ ENDIF(UNIX)
 
+# Enable the built-in updater by default. Linux packagers will find it useful to disable it
+# since package managers are responsible for updates there
+ADD_DEFINITIONS(-DINCLUDE_UPDATER=true)
```
Do you need to include a value - the presence of the `#define` is sufficient to trigger `#if defined(INCLUDE_UPDATER)` code - unless you plan to actually _test_ the value as well with `#if defined(INCLUDE_UPDATER) && (INCLUDE_UPDATER)` {if that _is_ the right way to do it... 😕}, and I get the feeling from a bit of searching that you should use `TRUE` not `true`...

SlySven commented on src/mudlet.pro:
```diff
@@ -221,6 +221,10 @@ macx:LIBS += \
 
+# Enable the built-in updater by default. Linux packagers will find it useful to disable it
+# since package managers are responsible for updates there
+DEFINES += INCLUDE_UPDATER=true
```
Same as my comment for the `CMakeLists.txt` that you do not need a value but that if you do it should be `TRUE` (or `1` as it happens) not `true`...

vadi2 commented:
150 commits and a month later, it builds on both macOS and Linux, using both qmake and cmake.

The macOS stuff still needs to be signed properly by Travis though.

keneanung commented:
This will need a push to the main repo for testing now.

vadi2 commented:
Right, thanks for the reminder.

vadi2 commented:
`: no identity found` ehh

vadi2 commented:
Very nice. The signed updates from Travis work.

vadi2 commented:
@SlySven @keneanung Linux and macOS update are working at a good stage now. I'd like to get this reviewed and merged, then I'll do the Windows one that builds on this.

keneanung commented on src/mudlet.pro:
```diff
@@ -50,7 +52,7 @@ exists("../3rdparty/edbee-lib/edbee-lib/edbee-lib.pri") {
 # Set the current Mudlet Version, unfortunately the Qt documentation suggests
 # that only a #.#.# form without any other alphanumberic suffixes is required:
-VERSION = 3.5.0
+VERSION = 3.5.7
```
You may want to reset this value to 3.5.0 again

vadi2 replied:
Fixed

keneanung commented on src/mudlet.pro:
```diff
@@ -461,6 +468,36 @@ macx: {
     ICON = icons/osx.icns
+
+    # possibly unnecessary
+#    QMAKE_LFLAGS += -Wl,-rpath,@loader_path/../Frameworks
```
If it showed that this was unnecessary indeed, can you remove it again?

vadi2 replied:
Removed

vadi2 commented on CI/travis.osx.after_success.sh:
```diff
@@ -4,20 +4,40 @@ set -e
 # we deploy only qmake and clang combination for macOS
 if [ "${Q_OR_C_MAKE}" = "qmake" ] && [ "${CC}" = "clang" ]; then
-  git clone https://github.com/Mudlet/installers.git "${TRAVIS_BUILD_DIR}/../installers"
+  git clone https://github.com/Mudlet/installers.git -b change-signing-name "${TRAVIS_BUILD_DIR}/../installers"
```
Fixed

vadi2 commented on src/CMakeLists.txt:
```diff
@@ -455,6 +464,11 @@ ENDIF()
+  target_link_libraries(mudlet
+          "-framework Sparkle"
+          "-framework AppKit"
+          "-F${CMAKE_CURRENT_SOURCE_DIR}/../3rdparty/sparkle")
```
It looks like one to me already - what do you mean?

keneanung replied:
I mean joining this last one with the 2 calls above (lines 459 and 458 which are not part of this snippet)

vadi2 commented:
I've removed the embedded Sparkle framework and added it as a CocoaPod instead. This removed over 200 additional files from the commit while introducing a dependency for yet-another-package-manager on macOS.

SlySven commented:
264 commits and rising - this PR needs something for _that_ - how about an __epic__ label. 😆

vadi2 commented:
Yeah it's definitely consumed my life for the last little while. Though
high the number of commits is partly explained by me coding on Ubuntu and
using git to transfer changes over to macOS for compiling. The mac keyboard
is just uncomfortable to use.

On Sat, 11 Nov 2017, 12:48 pm Stephen Lyons, <notifications@github.com>
wrote:

> 264 commits and rising - this PR needs something for *that* - how about
> an *epic* label. 😆
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/1318#issuecomment-343659239>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGxjLZ8iVEc3so-q7MA6qWlHcagYbigks5s1YmKgaJpZM4P0J55>
> .
>

vadi2 commented:
@SlySven @keneanung this is good for review now. I'd like to have it in the next update

vadi2 commented:
I found a problem: AppImage doesn't get updated properly. I was testing with running Mudlet from the IDE instead of the AppImage and didn't catch it.

vadi2 commented:
Fixed that bug.

SlySven commented on src/mudlet.pro:
```diff
@@ -219,6 +221,10 @@ macx:LIBS += \
 
+# Enable the built-in updater by default. Linux packagers will find it useful to disable it
+# since package managers are responsible for updates there
+DEFINES += INCLUDE_UPDATER=TRUE
```
I do not think you should assign a value to `INCLUDE_UPDATER` it may lead to coding errors where a coder looks for the value to be `TRUE` or `FALSE` when they should instead test for `defined(INCLUDE_UPDATER)` {which has the value of `1` if `INCLUDE_UPDATER` is defined and `0` otherwise}.

Besides which I am not exactly sure what `INCLUDE_UPDATER` will be replaced by in the preprocessor - all the escaping I had to do on the previously line to get the `LUA_DEFAULT_DIR` ASCII string value from the environment into the `LUA_DEFAULT_PATH` defined value was necessary and I am not entirely sure what `TRUE` will actually come out as...

vadi2 replied:
So you're saying it should be `DEFINES += INCLUDE_UPDATER=` ? That will hardly be intuitive to look at. Seems to be it's a pick your poison sort of thing.

SlySven replied:
No, just `DEFINES += INCLUDE_UPDATER` then we just check for it being `#if defined(`...')' !

vadi2 commented:
`UserInterfaceState.xcuserstate` is part of the fake Xcode project needed for CocoaPods to work. I switched to CocoaPods as it saved us embedding compiled Sparkle source code and reduced the files in this PR by 200.

vadi2 commented on src/main.cpp:
```diff
@@ -189,6 +193,13 @@ int main(int argc, char* argv[])
     QApplication* app = qobject_cast<QApplication*>(initApp.data());
 
+#if defined(Q_OS_WIN)
+    auto abortLaunch = runUpdate();
```
Moved.

Not sure about forcing to make the command line arguments work on Windows though. If you want to check Mudlets version, you don't have to use the command line - it is embedded in the binary and viewable in the Windows file explorer.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -441,12 +444,25 @@ mudlet::mudlet()
     connect(mactionMultiView, SIGNAL(triggered()), this, SLOT(slot_multi_view()));
     connect(mactionCloseProfile, SIGNAL(triggered()), this, SLOT(slot_close_profile()));
 
+    settings = getQSettings();
```
Renamed.

As for a `delete` - well, other similar pointers in `mudlet` don't have them. Does it matter that much anyway? If `mudlet` class is deleted, the application is shut down.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -441,12 +444,25 @@ mudlet::mudlet()
+#if defined(INCLUDE_UPDATER)
+    updater = new Updater(this, settings);
+#if defined(Q_OS_MACOS)
+    connect(dactionUpdate, &QAction::triggered, this, &mudlet::slot_check_manual_update);
```
Moved.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -2059,30 +2087,46 @@ void mudlet::closeEvent(QCloseEvent* event)
+    // pass the event on so dblsqd can perform an update
+    // if automatic updates have been disabled
     event->accept();
-    qApp->quit();
```
Yeah. It also prevents the updater from firing off the "last window is closed" to show the update dialog because of the explicit and redundant quit call.

SlySven replied:
OK.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -3301,3 +3340,46 @@ QString mudlet::getMudletPath(const mudletPathType mode, const QString& extra1,
+// returns whenever this is a release or development build of Mudlet
+bool mudlet::onDevelopmentVersion()
+{
+    auto devSuffix = QByteArray(APP_BUILD).trimmed();
```
Less macros, the better.

vadi2 replied:
Maybe this could be a `constrexpr` function.

SlySven replied:
If I know what that was (reading about [constexpr](http://en.cppreference.com/w/cpp/language/constexpr) on en.cppreference.com didn't really help me much, I'd better check Wikipedia) - so the best response to that I can give is _perhaps_ - hang-on this sounds like something @ahmedcharles would know... 😕 😜

vadi2 commented on src/updater.cpp:
```diff
@@ -0,0 +1,364 @@
+Updater::Updater(QObject* parent, QSettings* settings) : QObject(parent), mUpdateInstalled(false)
+{
+    Q_ASSERT_X(settings, "updater", "QSettings object is required for the updater to work");
+    this->settings = settings;
```
Fixed.

vadi2 commented:
@SlySven feedback applied, have a look. Check out if you can understand how the new `Updater` class works - it is difficult to juggle code re-usability and very different updating workflows on all 3 platforms, so I hope someone other than the current me can make sense of it.

vadi2 commented:
Building for both with and without the option would cost too much time and
effort. We'll build with the option on only.

Nobody will mind if Mudlet is bigger by 0.5mb, don't worry about it. We've
got bigger things to worry about!

On Tue, 14 Nov 2017, 6:04 pm Stephen Lyons, <notifications@github.com>
wrote:

> Ideally we ought to extend the CI build matrix to cover the compilation
> both with and without the updater code included - can I just check that the
> ability to include/exclude it is across all three primary build targets? I
> get the fun you must be having with the mental hoops needed to cover all
> three at once, for me tit was when I contemplated the option to both
> *include*/*exclude* binary translation files (around 0.5MBytes each
> though they get considerable smaller with the -compress option to lrelease)
> from the resource system (fortunately the QTranslator::load methods will
> handle either type in resource files...)
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/1318#issuecomment-344325839>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGxjGYyYDgauqmwCfErsDoI4dNRFC_Sks5s2ccygaJpZM4P0J55>
> .
>

SlySven commented:
> Nobody will mind if Mudlet is bigger by 0.5mb

That was 0.5MByte **per translation file** so that gets multiplied by the number of translations - which might become uncomfortable - but that is a worse case so it will probably not be so much an issue as it seems. :pray:🤞

vadi2 commented:
@SlySven I've addressed the last review comment, have a look again.

SlySven commented:
Humm, I am seeing two new git submodules `3rdparty/dblsqd` and `3rdparty/sparkle-glue` which are not currently in my local repository but surely they are NOT required for all platform/situations are they?

Also, because `git submodule update --init` will pull in all missing sub-modules that means that the existing code to fixup if the edbee-widget submodule is not present will (if it is _tickled_ by nuking the `3rdpart/edbee-lib` branch of my build environment to trigger it) also go and grab all of them.  While this may be nice I do not think we want/need to get the `3rdparty/sparkle-glue` on Linux platforms or even `3rdparty/dblsqd` for those (package builders and me on a normal day) who will NOT be defining `INCLUDE_UPDATER`...

I'm going to have a little play with this to see if I can improve on making the sub-module inclusion a little more controllable!

SlySven commented:
🗒 we will need to update our third party licence stuff:
+ for Sparkle it has an MIT main licence (+ a 2-clause BSD for some of the components it includes).
+ for the _Sparkle-glue_ stuff we need to find out what licence Remko Tronçon will allow his "Mixing Cocoa and Qt" to be used under - the default for a work that only carries a © marking is __*All Rights Reversed*__ if I recall correctly which is really not going to work for us. 😱 
+ for the DBLSQD SDK code we include it has an [Apache-2.0 licence](https://opensource.org/licenses/Apache-2.0)

The latter means we have at least include:
> Copyright [yyyy] [name of copyright owner]
>
>  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
>
>  http://www.apache.org/licenses/LICENSE-2.0
>
>  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

vadi2 commented:
Someone could be using the same source code repo across various OS's and it would be pretty annoying for them if things just didn't work across OS's. It would also make the whole thing more complicated if it conditionally checked out submodules depending on various conditions - let's keep it simple and maintainable, the disk space savings will be negligible :\

vadi2 commented:
I've added the licenses. `sparkle-glue` is released under BSD per https://el-tramo.be/blog/mixing-cocoa-and-qt/.

SlySven commented:
> I've added the licenses. sparkle-glue is released under BSD per https://el-tramo.be/blog/mixing-cocoa-and-qt/.

I am not seeing that licence type asserted there - where do you see it - perusing the source tarball indicated not even the `readme.md` actual says anything about terms of use and as I say, without something else "_All rights are reserved_" ... 😟

vadi2 commented:
He states it in the comments when someone asks him about it.

On Sun, 19 Nov 2017, 10:58 pm Stephen Lyons, <notifications@github.com>
wrote:

> I've added the licenses. sparkle-glue is released under BSD per
> https://el-tramo.be/blog/mixing-cocoa-and-qt/.
>
> I am not seeing that licence type asserted there - where do you see it -
> perusing the source tarball indicated not even the readme.md actual says
> anything about terms of use and as I say, without something else "*All
> rights are reserved*" ... 😟
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/1318#issuecomment-345553310>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGxjOXBqfDciGCwEl5DcOihVXyI4S7Fks5s4KSbgaJpZM4P0J55>
> .
>

vadi2 commented:
![selection_246](https://user-images.githubusercontent.com/110988/33001195-9d4a731e-cdac-11e7-8f2e-2a817ff3a7bf.png)

SlySven commented:
That is good enough for me - hopefully it'll keep @csmall happy (the Debian guys and gals are fairly thorough on licensing details IIRC).

SlySven commented:
Got some _improvements_ for this PR as https://github.com/vadi2/Mudlet/pull/8 ...! 😁

vadi2 commented:
Cool, thanks, I will have a look!

vadi2 commented:
Included

vadi2 commented:
The macOS build is now broken, mind taking a look?

SlySven commented:
Humm, the error occurs on the Qt 5.9 QMake build (obviously using the Clang/LLVM compiler) build only at the linker stage - I am not sure that it is anything I have done:

    ld: warning: directory not found for option '-F/Users/travis/build/Mudlet/Mudlet/3rdparty/sparkle-glue/../sparkle'
    ld: warning: object file (/usr/local/Cellar/qt/5.9.2/lib/libQt5UiTools.a(quiloader.o)) was built for newer OSX version (10.10) than being linked (10.7)
    ld: warning: object file (/usr/local/Cellar/qt/5.9.2/lib/libQt5UiTools.a(formbuilder.o)) was built for newer OSX version (10.10) than being linked (10.7)
    ld: warning: object file (/usr/local/Cellar/qt/5.9.2/lib/libQt5UiTools.a(textbuilder.o)) was built for newer  OSX version (10.10) than being linked (10.7)
    ld: warning: object file (/usr/local/Cellar/qt/5.9.2/lib/libQt5UiTools.a(formbuilderextra.o)) was built for newer OSX version (10.10) than being linked (10.7)
    ld: warning: object file (/usr/local/Cellar/qt/5.9.2/lib/libQt5UiTools.a(abstractformbuilder.o)) was built for newer OSX version (10.10) than being linked (10.7)
    ld: warning: object file (/usr/local/Cellar/qt/5.9.2/lib/libQt5UiTools.a(moc_quiloader.o)) was built for newer OSX version (10.10) than being linked (10.7)
    ld: warning: object file (/usr/local/Cellar/qt/5.9.2/lib/libQt5UiTools.a(ui4.o)) was built for newer OSX version (10.10) than being linked (10.7)
    ld: warning: object file (/usr/local/Cellar/qt/5.9.2/lib/libQt5UiTools.a(resourcebuilder.o)) was built for newer OSX version (10.10) than being linked (10.7)
    ld: warning: object file (/usr/local/Cellar/qt/5.9.2/lib/libQt5UiTools.a(properties.o)) was built for newer OSX version (10.10) than being linked (10.7)
    ld: warning: object file (/usr/local/Cellar/qt/5.9.2/lib/libQt5UiTools.a(moc_properties_p.o)) was built for newer OSX version (10.10) than being linked (10.7)
    duplicate symbol _main in:
        main.o
    duplicate symbol __Z17createApplicationRiPPcRj in:
        main.o
    duplicate symbol _spDebugConsole in:
        main.o
    duplicate symbol __Z8copyFontRK7QStringS1_ in:
        main.o
    duplicate symbol __ZN11AutoUpdaterD2Ev in:
        AutoUpdater.o
    duplicate symbol __ZN11AutoUpdaterD1Ev in:
        AutoUpdater.o
    duplicate symbol __ZN11AutoUpdaterD0Ev in:
        AutoUpdater.o
    duplicate symbol __ZTV11AutoUpdater in:
        AutoUpdater.o
    duplicate symbol __ZTS11AutoUpdater in:
        AutoUpdater.o
    duplicate symbol __ZTI11AutoUpdater in:
        AutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdater29automaticallyDownloadsUpdatesEv in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdater29automaticallyChecksForUpdatesEv in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdater15checkForUpdatesEv in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdater15setRelaunchFlagEv in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdater11justUpdatedEv in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdaterD2Ev in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdaterD1Ev in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdaterD0Ev in:
        SparkleAutoUpdater.o
    duplicate symbol __ZTV18SparkleAutoUpdater in:
        SparkleAutoUpdater.o
    duplicate symbol __ZTS18SparkleAutoUpdater in:
        SparkleAutoUpdater.o
    duplicate symbol __ZTI18SparkleAutoUpdater in:
        SparkleAutoUpdater.o
    duplicate symbol _OBJC_IVAR_$_SparkleDelegate.delegateHandler in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdaterC2ERK7QString in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdaterC1ERK7QString in:
        SparkleAutoUpdater.o
    duplicate symbol _OBJC_CLASS_$_SparkleDelegate in:
        SparkleAutoUpdater.o
    duplicate symbol _OBJC_METACLASS_$_SparkleDelegate in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdater32setAutomaticallyDownloadsUpdatesEb in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN18SparkleAutoUpdater32setAutomaticallyChecksForUpdatesEb in:
        SparkleAutoUpdater.o
    duplicate symbol __ZN16CocoaInitializerD2Ev in:
        CocoaInitializer.o
    duplicate symbol __ZN16CocoaInitializerC2Ev in:
        CocoaInitializer.o
    duplicate symbol __ZN16CocoaInitializerD1Ev in:
        CocoaInitializer.o
    duplicate symbol __ZN16CocoaInitializerC1Ev in:
        CocoaInitializer.o
    ld: 32 duplicate symbols for architecture x86_64
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    make: *** [Mudlet.app/Contents/MacOS/Mudlet] Error 1
    
    The command "if [ "${Q_OR_C_MAKE}" = "qmake" ]; then qmake -v; qmake ${SPEC} ../src/mudlet.pro && make -j2; else cmake --version; cmake .. && make -j2 && make check; fi" exited with 2.

I have kicked that build to try again, but this is outside of my skills area, let me see if it fails again...

SlySven commented:
Travis Announcement "We are updating to Xcode 8.3! By default, macOS builds will now use Xcode 8.3 unless otherwise specified." See [xcode8.3 OS X new default image on November 28th](https://blog.travis-ci.com/2017-11-21-xcode8-3-default-image-announce) which has been delayed by two weeks - I wonder if anything in the MacOS CI setup has changed already?

SlySven commented:
💡  Ah, ha - Both the Mudlet project and the Sparkle-Glue have `main` and `Autoupdate` class/`.cpp` files AND for QMake I've noticed in the past that all the project's files _including those from sub-modules_ are built in the same directory! (CMake in comparision builds sub-directories _in_ sub-directories...)

SlySven commented:
@vadi2 I've put up some information on Gitter about this but I think we should NOT be building the `mixing-cocoa-and-qt.pro` (which is an `TEMPLATE = app` project file producing a `main.o` object file which clashes with the Mudlet file of the same name) - instead I think we need a `sparkle-glue.pri` which contains a sub-set of the contents which can be included like the other sub-modules are!

vadi2 commented:
@SlySven I've narrowed it down to the changes you did around INCLUDE_UPDATER flag and submodule inclusion - see how https://github.com/Mudlet/Mudlet/pull/1318/commits/fc5b0d63223865209bb6131a486b9234676b2331 fails to build when that is included but when I revert it in https://github.com/Mudlet/Mudlet/pull/1318/commits/ecc5373cedd7c95797e48717976e960158b32312, it works fine. Could you fix this?

SlySven commented:
Could you clarify - how does the updating process work on __MacOS__? I am starting to get the feeling that it is a __separate__, _external_ executable in THAT case, is that true?

If it is, then the case in which the CI fails to build __is__ when the controllable inclusion of the updater is correct - in that it is trying to build both the separate executable and the main Mudlet one in the same directory and you have two separate applications both with `main.o` and `AutoUpdate.o` _in the same directory at the same time_ for the __qmake__ case!

I do not have a MacOS system so this is hard for me to assess but I suspect that we can use all of what I've put together (😀) plus for the MacOS + Updater case __only__ we'll have to add to the qmake `mudlet.pro`:

    CONFIG += object_parallel_to_source
    OBJECT_DIR = object

then try CI builds both with NO_UPDATER_INCLUDE set in the environment and without.  Incidentally one of those "Chatty" messages that you nuked was to help with determining whether the qmake (and the cmake) build systems had picked up on whether this had been detected - qmake displayed some really odd behaviour when I was putting that bit together and it turns out that it may not pick up things from the environment depending on the invoking shell when qmake is run.

vadi2 commented:
main.cpp from sparkle-glue is not necessary - it's just a demo app. We
don't have to build it

On Thu, 23 Nov 2017, 4:03 pm Stephen Lyons, <notifications@github.com>
wrote:

> Could you clarify - how does the updating process work on *MacOS*? I am
> starting to get the feeling that it is a *separate*, *external*
> executable in THAT case, is that true?
>
> If it is, then the case in which the CI fails to build *is* when the
> controllable inclusion of the updater is correct - in that it is trying to
> build both the separate executable and the main Mudlet one in the same
> directory and you have two separate applications both with main.o and
> AutoUpdate.o *in the same directory at the same time* for the *qmake*
> case!
>
> I do not have a MacOS system so this is hard for me to assess but I
> suspect that we can use all of what I've put together (😀) plus for the
> MacOS + Updater case *only* we'll have to add to the qmake mudlet.pro:
>
> CONFIG += object_parallel_to_source
> OBJECT_DIR = object
>
> then try CI builds both with NO_UPDATER_INCLUDE set in the environment and
> without. Incidentally one of those "Chatty" messages that you nuked was to
> help with determining whether the qmake (and the cmake) build systems had
> picked up on whether this had been detected - qmake displayed some really
> odd behaviour when I was putting that bit together and it turns out that it
> may not pick up things from the environment depending on the invoking shell
> when qmake is run.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/1318#issuecomment-346640484>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGxjH_QMwKQv9f8vlgLaZgrdusWZOveks5s5Yj_gaJpZM4P0J55>
> .
>

vadi2 commented:
I would like to have this in 3.6.0 this weekend.
